File: auth/src/main/java/com/diagorn/sparkathon/auth/service/JwtService.java
package com.diagorn.sparkathon.auth.service;

import com.diagorn.sparkathon.auth.config.properties.JwtProperties;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * JWT management service
 *
 * @author diagorn
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class JwtService {

    private final SecretKey SECRET_KEY = Keys.secretKeyFor(SignatureAlgorithm.HS256);

    private final JwtProperties jwtProperties;

    /**
     * Generate access token for user
     *
     * @param userDetails - user
     * @return access token
     */
    public String generateAccessToken(UserDetails userDetails) {
        Map<String, Object> claims = new HashMap<>();
        return doGenerateToken(claims, userDetails.getUsername(), jwtProperties.getAccessExpireTimeSec());
    }

    /**
     * Generate refresh token for user
     *
     * @param userDetails - user
     * @return refresh token
     */
    public String generateRefreshToken(UserDetails userDetails) {
        Map<String, Object> claims = new HashMap<>();
        return doGenerateToken(claims, userDetails.getUsername(), jwtProperties.getRefreshExpireTimeSec());
    }

    /**
     * Validate token
     *
     * @param token       - token to validate
     * @param userDetails - user
     * @return validity fact
     */
    public Boolean validateToken(String token, UserDetails userDetails) {
        var username = getUsernameFromToken(token);
        return (username.equals(userDetails.getUsername()) && !isTokenExpired(token));
    }

    /**
     * Get token expiration date
     *
     * @param token - token
     * @return expiration date
     */
    public Date getExpirationDateFromToken(String token) {
        return getClaimFromToken(token, Claims::getExpiration);
    }

    /**
     * Get username from token
     *
     * @param token - token
     * @return username
     */
    public String getUsernameFromToken(String token) {
        return getClaimFromToken(token, Claims::getSubject);
    }

    private Boolean isTokenExpired(String token) {
        final var expiration = getExpirationDateFromToken(token);
        return expiration.before(new Date());
    }

    private <T> T getClaimFromToken(String token, Function<Claims, T> claimResolver) {
        var claims = getAllClaimsFromToken(token);
        return claimResolver.apply(claims);
    }

    private Claims getAllClaimsFromToken(String token) {
        return Jwts.parserBuilder()
                .setSigningKey(SECRET_KEY)
                .build()
                .parseClaimsJws(token)
                .getBody();
    }

    private String doGenerateToken(Map<String, Object> claims, String subject, long expireSec) {
        return Jwts.builder()
                .setClaims(claims)
                .setSubject(subject)
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + expireSec * 1_00))
                .signWith(SECRET_KEY)
                .compact();
    }
}

File: auth/src/main/java/com/diagorn/sparkathon/auth/service/JwtService.kt
package com.diagorn.sparkathon.auth.service

import com.diagorn.sparkathon.auth.config.properties.JwtProperties
import com.diagorn.sparkathon.auth.utils.SECOND
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Service
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.function.Function

/**
 * JWT management service
 *
 * @author diagorn
 */
@Service
class JwtService(
    private val jwtProperties: JwtProperties
) {

    /**
     * Generate access token for user
     *
     * @param userDetails - user
     * @return access token
     */
    fun generateAccessToken(userDetails: UserDetails): String {
        val claims: Map<String, Any?> = emptyMap()
        return doGenerateToken(
            claims,
            userDetails.username,
            getRole(userDetails),
            jwtProperties.accessExpireTimeSec.toLong()
        )
    }

    /**
     * Generate refresh token for user
     *
     * @param userDetails - user
     * @return refresh token
     */
    fun generateRefreshToken(userDetails: UserDetails): String {
        val claims: Map<String, Any?> = HashMap()
        return doGenerateToken(
            claims,
            userDetails.username,
            getRole(userDetails),
            jwtProperties.refreshExpireTimeSec.toLong()
        )
    }

    /**
     * Validate token
     *
     * @param token       - token to validate
     * @param userDetails - user
     * @return validity fact
     */
    fun validateToken(token: String?, userDetails: UserDetails): Boolean {
        val username = try {
            getUsernameFromToken(token)
        } catch (e: Exception) {
            null
        }

        return username?.isNotEmpty() ?: false
                && username == userDetails.username
                && !isTokenExpired(token)
    }

    /**
     * Get token expiration date
     *
     * @param token - token
     * @return expiration date
     */
    fun getExpirationDateFromToken(token: String?): Date = getClaimFromToken(token) { obj: Claims -> obj.expiration }

    /**
     * Get username from token
     *
     * @param token - token
     * @return username
     */
    fun getUsernameFromToken(token: String?): String = getClaimFromToken(token) { obj: Claims -> obj.subject }

    private fun isTokenExpired(token: String?): Boolean {
        val expiration = try {
            getExpirationDateFromToken(token)
        } catch (e: Exception) {
            null
        }
        return expiration == null || expiration.before(Date())
    }

    private fun <T> getClaimFromToken(token: String?, claimResolver: Function<Claims, T>): T {
        val claims = getAllClaimsFromToken(token)
        return claimResolver.apply(claims)
    }

    private fun getAllClaimsFromToken(token: String?): Claims = Jwts.parserBuilder()
            .setSigningKey(Keys.hmacShaKeyFor(jwtProperties.secret.toByteArray(StandardCharsets.UTF_8)))
            .build()
            .parseClaimsJws(token)
            .body

    private fun doGenerateToken(claims: Map<String, Any?>, subject: String, role: String?, expireSec: Long): String =
        Jwts.builder()
            .setClaims(claims)
            .setSubject(subject)
            .claim(ROLE, role)
            .setIssuedAt(Date(System.currentTimeMillis()))
            .setExpiration(Date(System.currentTimeMillis() + expireSec * SECOND))
            .signWith(Keys.hmacShaKeyFor(jwtProperties.secret.toByteArray(StandardCharsets.UTF_8)))
            .compact()

    private fun getRole(userDetails: UserDetails): String? = userDetails.authorities.firstOrNull()?.authority

    companion object {
        const val ROLE = "role"
    }
}